README


This archive contains the distribution Dancer2-Plugin-Feed,
version 1.142060:

  Easy to generate feed rss or atom for Dancer2 applications.

This software is copyright (c) 2014 by Natal Ngétal.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.


This README file was generated by Dist::Zilla::Plugin::Readme v5.009.



Changes
Revision history for Dancer-Plugin-Feed

1.142060  2014-07-25 09:39:46CEST+0200 Europe/Paris

    * Use Plack::Test for all test thanks to sawyer (#2).

1.132820  2013-10-09 13:33:30 Europe/Paris

    * Initial version developed.



t/lib/TestApp.pm
#
# This file is part of Dancer2-Plugin-Feed
#
# This software is copyright (c) 2014 by Natal Ngétal.
#
# This is free software; you can redistribute it and/or modify it under
# the same terms as the Perl 5 programming language system itself.
#
package TestApp;

use Dancer2;
use Dancer2::Plugin::Feed;

get '/feed' => sub {
    create_feed(
        title   => 'this is a test',
        entries => [ { title => 'first entry' } ],
    );
};

get '/feed/:format' => sub {
    create_feed(
        format  => params->{format},
        title   => 'TestApp with ' . params->{format},
        entries => _get_entries(),
    );
};

get '/other/feed/rss' => sub {
    create_rss_feed(
        title   => 'TestApp with rss',
        entries => _get_entries(),
    );
};

get '/other/feed/atom' => sub {
    create_atom_feed(
        title   => 'TestApp with atom',
        entries => _get_entries(),
    );
};

sub _get_entries {
    [map { { title => "entry $_" } } ( 1 .. 10 )];
}

1;


Build.PL

# This file was automatically generated by Dist::Zilla::Plugin::ModuleBuild v5.009.
use strict;
use warnings;

use Module::Build 0.3601;


my %module_build_args = (
  "build_requires" => {
    "Module::Build" => "0.3601"
  },
  "configure_requires" => {
    "ExtUtils::MakeMaker" => "6.30",
    "Module::Build" => "0.3601"
  },
  "dist_abstract" => "Easy to generate feed rss or atom for Dancer2 applications.",
  "dist_author" => [
    "Natal Ng\x{e9}tal"
  ],
  "dist_name" => "Dancer2-Plugin-Feed",
  "dist_version" => "1.142060",
  "license" => "perl",
  "module_name" => "Dancer2::Plugin::Feed",
  "recommends" => {},
  "recursive_test_files" => 1,
  "requires" => {
    "Dancer" => "1.2001",
    "Dancer2" => 0,
    "Dancer2::Plugin" => 0,
    "XML::Feed" => 0
  },
  "script_files" => [],
  "test_requires" => {
    "HTTP::Request::Common" => 0,
    "Plack::Test" => 0,
    "Test::More" => 0,
    "lib" => 0,
    "strict" => 0,
    "warnings" => 0
  }
);


my %fallback_build_requires = (
  "HTTP::Request::Common" => 0,
  "Module::Build" => "0.3601",
  "Plack::Test" => 0,
  "Test::More" => 0,
  "lib" => 0,
  "strict" => 0,
  "warnings" => 0
);


unless ( eval { Module::Build->VERSION(0.4004) } ) {
  delete $module_build_args{test_requires};
  $module_build_args{build_requires} = \%fallback_build_requires;
}

my $build = Module::Build->new(%module_build_args);

$build->create_build_script;


lib/Dancer2/Plugin/Feed.pm
#
# This file is part of Dancer2-Plugin-Feed
#
# This software is copyright (c) 2014 by Natal Ngétal.
#
# This is free software; you can redistribute it and/or modify it under
# the same terms as the Perl 5 programming language system itself.
#
package Dancer2::Plugin::Feed;
{
  $Dancer2::Plugin::Feed::VERSION = '1.142060';
}

use Dancer2;
use Dancer2::Plugin;
use XML::Feed;

#ABSTRACT: Easy to generate feed rss or atom for Dancer2 applications.

my $ct = {
    atom => 'application/atom+xml',
    rss  => 'application/rss+xml',
};

my @feed_properties =
  qw/format title base link tagline description author id language copyright self_link modified/;

my @entries_properties =
  qw/title base link content summary category tags author id issued modified enclosure/;

register create_feed => sub {
    my ($dsl, %params) = @_;

    my $format = _validate_format(\%params);

    if ($format =~ /^atom$/i) {
        _create_atom_feed($dsl, \%params);
    }
    elsif($format =~/^rss$/i) {
        _create_rss_feed($dsl, \%params);
    }
    else {
        die "Unknown format $format, use rss or atom\n";
    }
};

register create_atom_feed => sub {
    my ($dsl, %params) = plugin_args(@_);

    _create_atom_feed($dsl, \%params);
};

register create_rss_feed => sub {
    my ($dsl, %params) = plugin_args(@_);

    _create_rss_feed($dsl, \%params);
};

sub _validate_format {
    my( $params, $dsl ) = @_;
    my $format = delete $params->{format};

    if (!$format) {
        my $settings = plugin_setting;
        $format      = $settings->{format}
            or die "Feed format is missing\n";
    }

    if ($format !~ /^(?:atom|rss)$/i) {
        die "Unknown format $format, use rss or atom\n";
    }

    return $format;
}

sub _create_feed {
    my ($format, $params) = @_;

    my $entries = delete $params->{entries};

    my $feed     = XML::Feed->new($format);
    my $settings = plugin_setting;

    map {
        my $val = $params->{$_} || $settings->{$_};
        $feed->$_($val) if ($val);
    } @feed_properties;

    foreach my $entry (@$entries) {
        my $e = XML::Feed::Entry->new($format);

        map {
            my $val = $entry->{$_};
            $e->$_($val) if $val
        } @entries_properties;

        $feed->add_entry($e);
    }

    return $feed->as_xml;
}

sub _create_atom_feed {
    my ($dsl, $params) = @_;

    $dsl->content_type($ct->{atom});
    _create_feed('Atom', $params);
}

sub _create_rss_feed {
    my ($dsl, $params) = @_;

    $dsl->content_type($ct->{rss});
    _create_feed('RSS', $params);
}

register_plugin for_versions => [2];

1;

__END__

=pod

=head1 NAME

Dancer2::Plugin::Feed - Easy to generate feed rss or atom for Dancer2 applications.

=head1 VERSION

version 1.142060

=head1 SYNOPSIS

    use Dancer2;
    use Dancer2::Plugin::Feed;
    use Try::Tiny;

    get '/feed/:format' => sub {
        my $feed;
        try {
            $feed = create_feed(
                format  => params->{format},
                title   => 'my great feed',
                entries => [ map { title => "entry $_" }, 1 .. 10 ],
            );
        }
        catch {
            my ( $exception ) = @_;

            if ( $exception->does('FeedInvalidFormat') ) {
                return $exception->message;
            }
            elsif ( $exception->does('FeedNoFormat') ) {
                return $exception->message;
            }
            else {
                $exception->rethrow;
            }
        };

        return $feed;
    };

    dance;

=head1 DESCRIPTION

Provides an easy way to generate RSS or Atom feed. This module relies on L<XML::Feed>. Please, consult the documentation of L<XML::Feed> and L<XML::Feed::Entry>.

=encoding UTF-8

=head1 CONFIGURATION

 plugins:
   Feed:
     title: my great feed
     format: Atom

=head1 FUNCTIONS

=head2 create_feed

This function returns a XML feed. All parameters can be define in the configuration

AcceptEd parameters are:

=over 4

=item format (required)

The B<Content-Type> header will be set to the appropriate value

=item entries

An arrayref containing a list of entries. Each item will be transformed to an L<XML::Feed::Entry> object. Each entry is an hashref. Some common attributes for these hashrefs are C<title>, C<link>, C<summary>, C<content>, C<author>, C<issued> and C<modified>. Check L<XML::Feed::Entry> for more details.

=item title

=item base

=item link

=item tagline

=item description

=item author

=item language

=item copyright

=item self_link

=item modified

=back

=head2 create_atom_feed

This method call B<create_feed> by setting the format to Atom.

=head2 create_rss_feed

This method call B<create_feed> by setting the format to RSS.

=head1 CONTRIBUTING

This module is developed on Github at:

L<http://github.com/hobbestigrou/Dancer2-Plugin-Feed>

Feel free to fork the repo and submit pull requests

=head1 ACKNOWLEDGEMENTS

Alexis Sukrieh and Franck Cuny

=head1 BUGS

Please report any bugs or feature requests in github.

=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc Dancer2::Plugin::Feed

=head1 SEE ALSO

L<Dancer2>
L<XML::Feed>
L<XML::Feed::Entry>
L<Dancer::Plugin::Feed>

=head1 AUTHOR

Natal Ngétal

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2014 by Natal Ngétal.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


Makefile.PL

# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v5.009.
use strict;
use warnings;



use ExtUtils::MakeMaker 6.30;



my %WriteMakefileArgs = (
  "ABSTRACT" => "Easy to generate feed rss or atom for Dancer2 applications.",
  "AUTHOR" => "Natal Ng\x{e9}tal",
  "BUILD_REQUIRES" => {
    "Module::Build" => "0.3601"
  },
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => "6.30",
    "Module::Build" => "0.3601"
  },
  "DISTNAME" => "Dancer2-Plugin-Feed",
  "EXE_FILES" => [],
  "LICENSE" => "perl",
  "NAME" => "Dancer2::Plugin::Feed",
  "PREREQ_PM" => {
    "Dancer" => "1.2001",
    "Dancer2" => 0,
    "Dancer2::Plugin" => 0,
    "XML::Feed" => 0
  },
  "TEST_REQUIRES" => {
    "HTTP::Request::Common" => 0,
    "Plack::Test" => 0,
    "Test::More" => 0,
    "lib" => 0,
    "strict" => 0,
    "warnings" => 0
  },
  "VERSION" => "1.142060",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


my %FallbackPrereqs = (
  "Dancer" => "1.2001",
  "Dancer2" => 0,
  "Dancer2::Plugin" => 0,
  "HTTP::Request::Common" => 0,
  "Module::Build" => "0.3601",
  "Plack::Test" => 0,
  "Test::More" => 0,
  "XML::Feed" => 0,
  "lib" => 0,
  "strict" => 0,
  "warnings" => 0
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);



